### PR TITLE
fix(lights): Add FLOOD LT MAIN PNL lights + new stepless potentiometer

### LINKED ISSUE #2429: [BUG] Integ Lights have small brightness range
**Mod Version**
  "built": "2020-12-14T10:21:06+00:00",
  "ref": "",
  "sha": "",
  "actor": "Harry <43104362+Snapmatics@users.noreply.github.com>",
  "event_name": "manual"

**Describe the bug**
When using the latest dev, with the new lighting features, all integ lights do not have correct brightnesses, flood lights work as they should, but integ lights become fully bright very quickly

i feel that they should fade in slower, or perhaps they do, and the differences are so small they are not noticable

**To Reproduce**
1. Load up acft
2. Power acft on gpu/engines
3.Turn off integ lights, then slowly move them to full brightness

**Expected behavior**
lights begin off, then progressively move to full brightness at 100% rotation of the knob

**Actual behavior**
lights begin off, then move to full brightness around 25-50% of the rotation

**References**
https://i.imgur.com/VM8J0Up.png - lighting at 100% brightness
https://i.imgur.com/9Z0fSbp.png - lighting after 3 clicks to the right, from 0%

**Additional context**

Was this working before/when did the issue start occurring?
after lighting PR was merged

Is this a problem in the vanilla unmodded game?
Not tested

Discord username (if different from GitHub):

### LINKED ISSUE #2901: [BUG] Flood lighting not accurate in main displays , PFD , ECAM ND AS THE KNOB FOR IT.
**Mod Version**
Latest branch

**Describe the bug**
The flood light in the main displays as i said before , PFD , ECAM , ND ETC is so inaccurate , the lowest setting for the knob in the display lights , just emits more light that it should be in real life , it should be a lot lower....

Also the knob for those (specially the knob for the main panel flood ), should be twist as the speed knob , right now its so inaccurate , its has kind of pre selected modes so it goes from nothing to so much

**To Reproduce**

1.Twist those buttons 
2.
3.

**Expected behavior**
Even lower flood light for the main displays , as in real life , i just tested , I am a Cabin crew you can check that on discord Mikecbr and the Knob for the MAIN DISPLAY PANEL should twist like the speed one.

**Actual behavior**
Not realistic , so much light even in the lowest mode for the main displays and the  Knob for the MAIN DISPLAY PANEL not twisting like it should be

**References**

**Additional context**

Was this working before/when did the issue start occurring?
No

Is this a problem in the vanilla unmodded game?
Dont think so

Discord username (if different from GitHub):

### PR DESCRIPTION
**Pull request by bouveng**
flood-lights-wip-alt-comp

Fixes #2901, #1450

## Summary of Changes

* Adds new FLOOD LT MAIN PNL lights with proper and realistic indirect light/ambience. 
* Adds a new stepless, silent, precision potentiometer for use with various lights.
* Lights have a great dimmable range and can be adjusted with clicks for precision setup or with longpress for rapid change.

## Screenshots

Microsoft Flight Simulator 2021-06-21 09_17_42

Microsoft Flight Simulator 2021-06-21 09_16_26

## Video example
https://user-images.githubusercontent.com/433429/122734980-df5c1c00-d27e-11eb-9005-51899a29d646.mp4

Demonstration of rapid change of intensity with longpress and regular clicks for precise setup.

## References

> They are just rotary knobs with no off indentation to confirm it's off.  Just spin it one way for light and the other for darkness 
— Tricker

19-cockpit-avion-airbusA320

memory3

download

## Testing instructions

1, Spawn C&D, somewhere dark, with as little external lights as possible (on RW at ESME for example)
2, Power aircraft in suitable way.
3, Turn of any interfering lights, integrated lights, screen, etc to your liking.
4, Try out the FLOOD LT MAIN PNL 
5, Get cozy and listen to song "OV Wright - I Don't Know Why"

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Review by Benjozork — APPROVED** [review]

**Comment by sidnov** [comment]
@bouveng "will fix" would actually close 2429 too. You might want to edit that 😄 
image

Great PR otherwise! 🚀

**Comment by bouveng** [comment]
> @bouveng "will fix" would actually close 2429 too. You might want to edit that 😄

What do you mean? This PR does nothing for INTEG LT. But the new stepless potentiometer will give the INTEG LT a greater "range" and will probably fix that issue.

**Comment by tracernz** [comment]
> > @bouveng "will fix" would actually close 2429 too. You might want to edit that 😄
> 
> What do you mean? This PR does nothing for INTEG LT. But the new stepless potentiometer will give the INTEG LT a greater "range" and will probably fix that issue.

Edit your comment to remove the word fix in front of those, otherwise GitHub will automatically close those issues as fixed by this PR when it merges.

**Comment by bouveng** [comment]
> > > @bouveng "will fix" would actually close 2429 too. You might want to edit that 😄
> > 
> > 
> > What do you mean? This PR does nothing for INTEG LT. But the new stepless potentiometer will give the INTEG LT a greater "range" and will probably fix that issue.
> 
> Edit your comment to remove the word fix in front of those, otherwise GitHub will automatically close those issues as fixed by this PR when it merges.

Ah. Right. Did not know that. 🙈

File: flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/systems.cfg [hunk 1]
```diff
@@ -73,32 +73,19 @@ lightdef.19 = Type:10#Index:2#LocalPosition:34.8,1.0,3.7#LocalRotation:270,500,0
...
+lightdef.23 = Type:12#Index:4#LocalPosition:38.97,0.7,6.2#LocalRotation:-10,0,0#EffectFile:LIGHT_A32NX_MainPanelFloodCenter#PotentiometerIndex:83   ; F/O L
+lightdef.24 = Type:12#Index:4#LocalPosition:38.97,2.6,5.87#LocalRotation:-10,0,8#EffectFile:LIGHT_A32NX_MainPanelFloodEnd#PotentiometerIndex:83     ; F/O R
+
+; Main panel flood lights ambient fx
+lightdef.25 = Type:12#Index:3#LocalPosition:38.8,-2.58,5.78#LocalRotation:0,0,0#EffectFile:LIGHT_A32NX_MainPanelFloodAmbientEnd#PotentiometerIndex:83   ; CPT
+lightdef.26 = Type:12#Index:3#LocalPosition:38.75,0,5.8#LocalRotation:0,0,0#EffectFile:LIGHT_A32NX_MainPanelFloodAmbientCenter#PotentiometerIndex:83    ; Center
+lightdef.27 = Type:12#Index:4#LocalPosition:38.8,2.58,5.78#LocalRotation:0,0,0#EffectFile:LIGHT_A32NX_MainPanelFloodAmbientEnd#PotentiometerIndex:83    ; F/O
+
+; lightdef.28 unused
+; lightdef.29 unused
```
Comment (davidwalschots) [inline flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/systems.cfg:88]: Last time I checked (at least this is true for circuits), skipping indexes doesn't work as anything after the skipped index basically isn't working.

File: flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/systems.cfg [hunk 2]
```diff
@@ -73,32 +73,19 @@ lightdef.19 = Type:10#Index:2#LocalPosition:34.8,1.0,3.7#LocalRotation:270,500,0
...
+; Main panel flood lights
+lightdef.21 = Type:12#Index:3#LocalPosition:38.97,-2.6,5.87#LocalRotation:-10,0,-8#EffectFile:LIGHT_A32NX_MainPanelFloodEnd#PotentiometerIndex:83   ; CPT L
+lightdef.22 = Type:12#Index:3#LocalPosition:38.97,-0.6,6.2#LocalRotation:-10,0,0#EffectFile:LIGHT_A32NX_MainPanelFloodCenter#PotentiometerIndex:83  ; CPT R
+lightdef.23 = Type:12#Index:4#LocalPosition:38.97,0.7,6.2#LocalRotation:-10,0,0#EffectFile:LIGHT_A32NX_MainPanelFloodCenter#PotentiometerIndex:83   ; F/O L
+lightdef.24 = Type:12#Index:4#LocalPosition:38.97,2.6,5.87#LocalRotation:-10,0,8#EffectFile:LIGHT_A32NX_MainPanelFloodEnd#PotentiometerIndex:83     ; F/O R
+
+; Main panel flood lights ambient fx
+lightdef.25 = Type:12#Index:3#LocalPosition:38.8,-2.58,5.78#LocalRotation:0,0,0#EffectFile:LIGHT_A32NX_MainPanelFloodAmbientEnd#PotentiometerIndex:83   ; CPT
+lightdef.26 = Type:12#Index:3#LocalPosition:38.75,0,5.8#LocalRotation:0,0,0#EffectFile:LIGHT_A32NX_MainPanelFloodAmbientCenter#PotentiometerIndex:83    ; Center
+lightdef.27 = Type:12#Index:4#LocalPosition:38.8,2.58,5.78#LocalRotation:0,0,0#EffectFile:LIGHT_A32NX_MainPanelFloodAmbientEnd#PotentiometerIndex:83    ; F/O
```
Comment (davidwalschots) [inline flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/systems.cfg:85]: How does this play out when certain buses are unpowered? At the moment it seems a situation in which the CPT side is unpowered will also lead to all the ambient light in the centre disappearing, while that maybe shouldn't be the case. Similarly, when the F/O side is out, the centre might have too much light.

**Comment by Snapmatics** [comment]
That's very cool. I have been struggling with this one. Tag me when it's merged so I can do my light changes on top of this one.

File: flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/systems.cfg [hunk 3]
```diff
@@ -73,32 +73,19 @@ lightdef.19 = Type:10#Index:2#LocalPosition:34.8,1.0,3.7#LocalRotation:270,500,0
...
-lightdef.26 = Type:12#Index:3#LocalPosition:37.97,-2.45,4.87#LocalRotation:0,0,0#EffectFile:LIGHT_A32NX_CockpitMinimalAmbiantLow#PotentiometerIndex:83
-
-; Captain right glareshield (fake) ambient light
-lightdef.27 = Type:12#Index:3#LocalPosition:38,-0.6,4.8#LocalRotation:0,0,0#EffectFile:LIGHT_A32NX_CockpitMinimalAmbiantLow#PotentiometerIndex:83
-
-; F/O left glareshield (fake) ambient light
-lightdef.28 = Type:12#Index:4#LocalPosition:38,0.7,4.8#LocalRotation:0,0,0#EffectFile:LIGHT_A32NX_CockpitMinimalAmbiantLow#PotentiometerIndex:83
-
-; F/O right glareshield (fake) ambient light
-lightdef.29 = Type:12#Index:4#LocalPosition:37.97,2.45,4.87#LocalRotation:0,0,0#EffectFile:LIGHT_A32NX_CockpitMinimalAmbiantLow#PotentiometerIndex:83
```
Comment (davidwalschots) [inline flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/systems.cfg:101]: **issue**: Unused effect file still in repository.
**suggestion**: Remove `LIGHT_A32NX_CockpitMinimalAmbiantLow.fx`.

**Review by davidwalschots — CHANGES_REQUESTED** [review]

File: flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/systems.cfg (hunk 1 above)
Comment (bouveng) [reply flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/systems.cfg:88]: Hi @davidwalschots 

I have not encountered any issues with this. This indexing, at least when it comes to lights, _seems_ just to be an internal one as sim re-arranges these in the simulation anyway.

In attached image you can se that both `lightdef.28` and `lightdef.29` gets populated with the next definitions in `systems.cfg`

I can't find any documentation from MS or P3D that states that lightdefinitions needs to be in a particial order.

I could re-arrange them to be in sequence, but it makes little sense?

lightdefN

File: flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/systems.cfg (hunk 2 above)
Comment (bouveng) [reply flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/systems.cfg:85]: Hi @davidwalschots, I do not understand, all these lights are on the same type (`Type:12`)? Isn't that the bus?
I know you have been into this file and done work here earlier. Enlighten me please and/or suggest a change.

` ; light.n = CircuitType, ZOffset, XOffset, YOffset, XRotationInDegrees, YRotationInDegrees, ZRotationInDegrees, fxfile`

File: flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/systems.cfg (hunk 3 above)
Comment (bouveng) [reply flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/systems.cfg:101]: File removed.

**Comment by bouveng** [comment]
> Tag me when it's merged so I can do my light changes on top of this one.

Hi H! What _light changes_ are you thinking about? The lights in this PR have been changed in terms of `position`, `angle`, `rotation`, `range`, `intentsity`, `color `and `spot` shape, in order to work well with the 3 ambient lights positioned to make a realistic ambient/bounce.

Just curious what things you suggest be changed/modified? You should check this out in a dark cockpit, this is extremely beautiful and realistic looking.

File: flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/systems.cfg (hunk 2 above)
Comment (davidwalschots) [reply flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/systems.cfg:85]: No, Type:12 is not equal to bus 12. From the documentation I wrote within the `systems.cfg` file:

>; The lightdef.Type integer and circuit.Type enum are related. Below we list the
; known relationships between these values.
;
; lightdef.Type = circuit.Type
; 1 = CIRCUIT_LIGHT_BEACON
; 2 = CIRCUIT_LIGHT_STROBE
; 3 = CIRCUIT_LIGHT_NAV
; 4 = CIRCUIT_LIGHT_PANEL
; 5 = CIRCUIT_LIGHT_LANDING
; 6 = CIRCUIT_LIGHT_TAXI
; 7 = CIRCUIT_LIGHT_RECOGNITION
; 8 = CIRCUIT_LIGHT_WING
; 9 = CIRCUIT_LIGHT_LOGO
; 10 = CIRCUIT_LIGHT_CABIN,
; 11 = CIRCUIT_LIGHT_PEDESTAL
; 12 = CIRCUIT_LIGHT_GLARESHIELD
;
; You will find other types being used, but it is unclear how they relate. For example,
; lightdef.Type 13 seems to be always on, and can only be brightened by adapting the associated potentiometer.
;
; lightdef.Index and circuit.Type are related. When the lightdef.Index and circuit.Type numeric suffix match,
; then that circuit affects the powered state of the light. For example, these lightdef and circuit definitions
; are related:
; lightdef.35 = Type:4#Index:5
; circuit.81 = Type:CIRCUIT_LIGHT_PANEL:5#Connections:bus.7
;
; When electrical bus 7 is powered, then circuit 81 is powered, which in turn can power lightdef 35.

Thus, all lights with `Type:12#Index:3` are powered by `circuit.60 = Type:CIRCUIT_LIGHT_GLARESHIELD:3`; and all lights with `Type:12#Index:4` are powered by `circuit.82 = Type:CIRCUIT_LIGHT_GLARESHIELD:4`.

Thus, when DC1 is unpowered, `lightdef.23`, `lightdef.24`, and `lightdef.27` won't illuminate. When DC ESS is unpowered, `lightdef.21`, `lightdef.22`, `lightdef.25`, and `lightdef.26` won't illuminate.

You can test situations in which theses buses (but not the other) become unpowered by:

**DC ESS without DC1**
Fly and turn off both ENG GENs (while APU GEN is not available either). AC1, AC2 and therefore DC1 and DC2 will be unavailable. DC ESS is powered initially by the batteries and later by the EMER GEN.

**DC1 without DC ESS**
- Push the BUS TIE to OFF.
- Push AC ESS FEED to ALTN.
- Push GEN2 to OFF.

Results in AC1, DC1, DC BAT and DC2 being powered, while AC 2, AC ESS, and DC ESS become unpowered.

Comment (bouveng) [reply flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/systems.cfg:85]: @davidwalschots Thanks for valuable insight and explanation. Will seek ground truth, and make sure they are powered correctly.

**Comment by Snapmatics** [comment]
> > Tag me when it's merged so I can do my light changes on top of this one.
> 
> Hi @Snapmatics! What _light changes_ are you thinking about? The lights in this PR have been changed in terms of `position`, `angle`, `rotation`, `range`, `intentsity`, `color `and `spot` shape, in order to work well with the 3 ambient lights positioned to make a realistic ambient/bounce.
> 
> Just curious what things you suggest be changed/modified? You should check this out in a dark cockpit, this is extremely beautiful and realistic looking.

Nope, dm me on discord.

**Comment by Benjozork** [comment]
Please have these discussions in public as much as possible. This is an open project.

**Review by davidwalschots — APPROVED** [review]

**Comment by St54Kevin** [comment]
**Quality Assurance Tester**

Object of testing: #5177 
Tier of Testing: 1
Date: 22/06/2021

**Testing Process:**
1. Loaded C&D in the night 
2. Tested various knob settings in different lighting conditions

**Negatives:**
- Mouse Wheel usage is pretty slow now, which I would personally prefere to use. Using the click spots works better for me

**Testing Results:** Passed

**Comment by St54Kevin** [comment]
I changed to "Not Passed" to be in line with my other reviews until we see what other QAs think of it

**Comment by bouveng** [comment]
> * Mouse Wheel usage is pretty slow now, which I would personally prefere to use. Using the click spots works better for me

Hi K!

Since the knob now has 100 "steps" instead of 10, I can imagine the mouse wheel experience becomes a bit hard.
I use my mouse-wheel to zoom so I never thought of this.

But, this must be a problem for all the other "stepless" dials, like the 3 temperature knobs on the overhead?
Do they work better with mouse-wheel?

If you know a dial or knob that behaves nice with mouse-wheel and have more than 10 steps, let mw know and I can investigate it.

I do not know how to split klicks and mouse-wheel interactions so we can get "bigger" steps when using mouse-wheel.
So I do not know how to resolve this mouse-wheel issue at the moment.

**Comment by davidwalschots** [comment]
I don't know about other users, but I primarily use the mouse wheel for these knobs as well. Maybe you could go from 10 to 20 or 25 steps as a trade-off?

**Comment by davidwalschots** [comment]
#1450 is related to this.

**Comment by bouveng** [comment]
> I don't know about other users, but I primarily use the mouse wheel for these knobs as well. 

Ok. How does the air conditioning knobs work with the mouse wheel?

> Maybe you could go from 10 to 20 or 25 steps as a trade-off?

That would take away a large part of the "stepless"-ness. Now it is 100 steps. 0-100%.
This makes the lower range so fine-grain configurable. I would say the lower end of these lights are most important from a night flying perspective.

I could try 50 steps. But the whole point is not to have any steps, and simulate the real potentiometer.

**Comment by bouveng** [comment]
**UPDATE!**

Hi testers!

I have been digging around in the codebase and I think I will be able to turn the knobs differently when using the mouse-wheel.

There is `WheelUp` and `WheelDown` events!

I will have to do my own template for this. That is a good idea anyway since this will be used in a few places.
Give me a couple of days to build this.
Expect updated PR towards/in the weekend.

**Comment by bouveng** [comment]
This PR has been **closed** and **replaced** by #5244

**Pull request closed without merging**